pkg/standalone-utils/test/LidoRelayer.test.ts

```diff
@@ -1,462 +1,950 @@
+          ).wait();
+
+          expectTransferEvent(
+            receipt,
+            { from: ZERO_ADDRESS, to: relayer.address, value: expectedWstETHAmount },
+            wstETH
+          );
         });
+      }
+    });
+  });
 
-        it('does not take wstETH from the sender', async () => {
-          const wstETHBalanceBefore = await wstETH.balanceOf(sender);
-          await relayer.connect(sender).joinPool(basePoolId, sender.address, sender.address, joinRequest);
+  describe('complex actions', () => {
+    describe('swap', () => {
+      function encodeSwap(params: {
+        poolId: string;
+        kind: SwapKind;
+        tokenIn: Token;
+        tokenOut: Token;
+        amount: BigNumberish;
+        sender: Account;
+        recipient: Account;
+        outputReference?: BigNumberish;
+      }): string {
+        return relayerLibrary.interface.encodeFunctionData('swap', [
+          {
+            poolId: params.poolId,
+            kind: params.kind,
+            assetIn: params.tokenIn.address,
+            assetOut: params.tokenOut.address,
+            amount: params.amount,
+            userData: '0x',
+          },
+          {
+            sender: TypesConverter.toAddress(params.sender),
+            recipient: TypesConverter.toAddress(params.recipient),
+            fromInternalBalance: false,
+            toInternalBalance: false,
+          },
+          0,
+          MAX_UINT256,
+          0,
+          params.outputReference ?? 0,
+        ]);
+      }
+
+      describe('swap using stETH as an input', () => {
+        it('performs the given swap', async () => {
+          const poolId = basePoolId;
+          const tokenIn = wstETH;
+          const tokenOut = stETH;
+          const amount = fp(1);
+
+          const receipt = await (
+            await relayer.connect(sender).multicall([
+              encodeWrap(sender.address, relayer.address, amount, toChainedReference(0)),
+              encodeApprove(tokenIn, MAX_UINT256),
+              encodeSwap({
+                poolId,
+                kind: SwapKind.GivenIn,
+                tokenIn,
+                tokenOut,
+                amount: toChainedReference(0),
+                sender: relayer,
+                recipient,
+                outputReference: 0,
+              }),
+            ])
+          ).wait();
+
+          expectEvent.inIndirectReceipt(receipt, vault.instance.interface, 'Swap', {
+            poolId,
+            tokenIn: tokenIn.address,
+            tokenOut: tokenOut.address,
+            // amountIn: singleSwap.amount,
+            // amountOut
+          });
 
-          const wstETHBalanceAfter = await wstETH.balanceOf(sender);
-          expect(wstETHBalanceAfter).to.be.eq(wstETHBalanceBefore);
+          expectEvent.inIndirectReceipt(
+            receipt,
+            new Interface((await getArtifact('v2-solidity-utils/ERC20')).abi),
+            'Transfer',
+            { to: recipient.address }
```

Comment:
Shouldn't we at least check that the transfer the recipient got was of the token out?

Reply:
Same in the other swap tests.